<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RIBO-Seq Workflow Template • systemPipeRIBOseq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="RIBO-Seq Workflow Template">
<meta property="og:description" content="systemPipeRIBOseq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-158796879-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158796879-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeRIBOseq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/systemPipeRIBOseq.html">
    <span class="fas fa fas fa-book"></span>
     
    Workflow
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/systemPipeRIBOseq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeRIBOseq_files/header-attrs-2.4/header-attrs.js"></script><script src="systemPipeRIBOseq_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RIBO-Seq Workflow Template</h1>
                        <h4 class="author">Author: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>) and Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>)</h4>
            
            <h4 class="date">Last update: 14 October, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/systemPipeR/systemPipeRIBOseq/blob/master/vignettes/systemPipeRIBOseq.Rmd"><code>vignettes/systemPipeRIBOseq.Rmd</code></a></small>
      <div class="hidden name"><code>systemPipeRIBOseq.Rmd</code></div>

    </div>

    
    
<!--
- Compile from command-line
Rscript -e "rmarkdown::render('systemPipeRIBOseq.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('systemPipeRIBOseq.Rmd', tangle=TRUE)"; Rscript -e "rmarkdown::render('systemPipeRIBOseq.Rmd', c('BiocStyle::pdf_document'))"
-->
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Ribo-Seq and polyRibo-Seq are a specific form of RNA-Seq gene expression experiments utilizing mRNA subpopulations directly bound to ribosomes. Compared to standard RNA-Seq, their readout of gene expression provides a better approximation of downstream protein abundance profiles due to their close association with translational processes. The most important difference among the two is that polyRibo-Seq utilizes polyribosomal RNA for sequencing, whereas Ribo-Seq is a footprinting approach restricted to sequencing RNA fragments protected by ribosomes <span class="citation">(Ingolia et al. 2009; Aspden et al. 2014; Juntawong et al. 2015)</span>.</p>
<p>The workflow presented in this vignette contains most of the data analysis steps described by <span class="citation">(Juntawong et al. 2014)</span> including functionalities useful for processing both polyRibo-Seq and Ribo-Seq experiments. To improve re-usability and adapt to recent changes of software versions (<em>e.g.</em> R, Bioconductor and short read aligners), the code has been optimized accordingly. Thus, the results obtained with the updated workflow are expected to be similar but not necessarily identical with the published results described in the original paper.</p>
<p>Relevant analysis steps of this workflow include read preprocessing, read alignments against a reference genome, counting of reads overlapping with a wide range of genomic features (<em>e.g.</em> CDSs, UTRs, uORFs, rRNAs, etc.), differential gene expression and differential ribosome binding analyses, as well as a variety of genome-wide summary plots for visualizing RNA expression trends. Functions are provided for evaluating the quality of Ribo-seq data, for identifying novel expressed regions in the genomes, and for gaining insights into gene regulation at the post-transcriptional and translational levels. For example, the functions <code>genFeatures</code> and <code>featuretypeCounts</code> can be used to quantify the expression output for all feature types included in a genome annotation (<code>e.g.</code> genes, introns, exons, miRNAs, intergenic regions, etc.). To determine the approximate read length of ribosome footprints in Ribo-Seq experiments, these feature type counts can be obtained and plotted for specific read lengths separately. Typically, the most abundant read length obtained for translated features corresponds to the approximate footprint length occupied by the ribosomes of a given organism group. Based on the results from several Ribo-Seq studies, these ribosome footprints are typically ~30 nucleotides long <span class="citation">(Ingolia, Lareau, and Weissman 2011; Ingolia et al. 2009; Juntawong et al. 2014)</span>. However, their length can vary by several nucleotides depending upon the optimization of the RNA digestion step and various factors associated with translational regulation. For quality control purposes of Ribo-Seq experiments it is also useful to monitor the abundance of reads mapping to rRNA genes due to the high rRNA content of ribosomes. This information can be generated with the <code>featuretypeCounts</code> function described above.</p>
<p>Coverage trends along transcripts summarized for any number of transcripts can be obtained and plotted with the functions <code>featureCoverage</code> and <code>plotfeatureCoverage</code>, respectively. Their results allow monitoring of the phasing of ribosome movements along triplets of coding sequences. Commonly, high quality data will display here for the first nucleotide of each codon the highest depth of coverage computed for the 5’ ends of the aligned reads.</p>
<p>Ribo-seq data can also be used to evaluate various aspects of translational control due to ribosome occupancy in upstream open reading frames (uORFs). The latter are frequently present in (or near) 5’ UTRs of transcripts. For this, the function <code>predORFs</code> can be used to identify ORFs in the nucleotide sequences of transcripts or their subcomponents such as UTR regions. After scaling the resulting ORF coordinates back to the corresponding genome locations using <code>scaleRanges</code>, one can use these novel features (<em>e.g.</em> uORFs) for expression analysis routines similar to those employed for pre-existing annotations, such as the exonic regions of genes. For instance, in Ribo-Seq experiments one can use this approach to systematically identify all transcripts occupied by ribosomes in their uORF regions. The binding of ribosomes to uORF regions may indicate a regulatory role in the translation of the downstream main ORFs and/or translation of the uORFs into functionally relevant peptides.</p>
<div id="experimental-design" class="section level2">
<h2 class="hasAnchor">
<a href="#experimental-design" class="anchor"></a>Experimental design</h2>
<p>Typically, users want to specify here all information relevant for the analysis of their NGS study. This includes detailed descriptions of FASTQ files, experimental design, reference genome, gene annotations, etc.</p>
</div>
</div>
<div id="workflow-environment" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-environment" class="anchor"></a>Workflow environment</h1>
<div id="load-packages-and-sample-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-packages-and-sample-data" class="anchor"></a>Load packages and sample data</h2>
<p>The <code>systemPipeR</code> package needs to be loaded to perform the analysis steps shown in this report <span class="citation">(H Backman and Girke 2016)</span>. The package allows users to run the entire analysis workflow interactively or with a single command while also generating the corresponding analysis report. For details see <code>systemPipeR's</code> main <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">vignette</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://girke.bioinformatics.ucr.edu/systemPipeR/">systemPipeR</a></span><span class="op">)</span></pre></div>
</div>
<div id="generate-workflow-environment" class="section level2">
<h2 class="hasAnchor">
<a href="#generate-workflow-environment" class="anchor"></a>Generate workflow environment</h2>
<p><a href="http://bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html"><em>systemPipeRdata</em></a> package is a helper package to generate a fully populated <a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html"><em>systemPipeR</em></a> workflow environment in the current working directory with a single command. All the instruction for generating the workflow template are provide in the <em>systemPipeRdata</em> vignette <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html#1_Introduction">here</a>.</p>
<p>After building and loading the workflow environment generated by <code>genWorkenvir</code> from <em>systemPipeRdata</em> all data inputs are stored in a <code>data/</code> directory and all analysis results will be written to a separate <code>results/</code> directory, while the <code>systemPipeRIBOseq.Rmd</code> script and the <code>targets</code> file are expected to be located in the parent directory. The R session is expected to run from this parent directory. Additional parameter files are stored under <code>param/</code>.</p>
<p>To work with real data, users want to organize their own data similarly and substitute all test data for their own data. To rerun an established workflow on new data, the initial <code>targets</code> file along with the corresponding FASTQ files are usually the only inputs the user needs to provide.</p>
<p>For more details, please consult the documentation <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">here</a>. More information about the <code>targets</code> files from <em>systemPipeR</em> can be found <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#25_structure_of_targets_file">here</a>.</p>
</div>
<div id="run-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#run-workflow" class="anchor"></a>Run workflow</h2>
<p>Now open the R markdown script <code>systemPipeRIBOseq.Rmd</code>in your R IDE (<em>e.g.</em> vim-r or RStudio) and run the workflow as outlined below.</p>
<p>Here pair-end workflow example is provided. Please refer to the main vignette <code>systemPipeR.Rmd</code> for running the workflow with single-end data.</p>
<p>If you are running on a single machine, use following code as an example to check if some tools used in this workflow are in your environment <strong>PATH</strong>. No warning message should be shown if all tools are installed.</p>
</div>
<div id="experiment-definition-provided-by-targets-file" class="section level2">
<h2 class="hasAnchor">
<a href="#experiment-definition-provided-by-targets-file" class="anchor"></a>Experiment definition provided by <code>targets</code> file</h2>
<p>The <code>targets</code> file defines all FASTQ files and sample comparisons of the analysis workflow.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="va">targets</span></pre></div>
<pre><code>##                      FileName1                   FileName2
## 1  ./data/SRR446027_1.fastq.gz ./data/SRR446027_2.fastq.gz
## 2  ./data/SRR446028_1.fastq.gz ./data/SRR446028_2.fastq.gz
## 3  ./data/SRR446029_1.fastq.gz ./data/SRR446029_2.fastq.gz
## 4  ./data/SRR446030_1.fastq.gz ./data/SRR446030_2.fastq.gz
## 5  ./data/SRR446031_1.fastq.gz ./data/SRR446031_2.fastq.gz
## 6  ./data/SRR446032_1.fastq.gz ./data/SRR446032_2.fastq.gz
## 7  ./data/SRR446033_1.fastq.gz ./data/SRR446033_2.fastq.gz
## 8  ./data/SRR446034_1.fastq.gz ./data/SRR446034_2.fastq.gz
## 9  ./data/SRR446035_1.fastq.gz ./data/SRR446035_2.fastq.gz
## 10 ./data/SRR446036_1.fastq.gz ./data/SRR446036_2.fastq.gz
## 11 ./data/SRR446037_1.fastq.gz ./data/SRR446037_2.fastq.gz
## 12 ./data/SRR446038_1.fastq.gz ./data/SRR446038_2.fastq.gz
## 13 ./data/SRR446039_1.fastq.gz ./data/SRR446039_2.fastq.gz
## 14 ./data/SRR446040_1.fastq.gz ./data/SRR446040_2.fastq.gz
## 15 ./data/SRR446041_1.fastq.gz ./data/SRR446041_2.fastq.gz
## 16 ./data/SRR446042_1.fastq.gz ./data/SRR446042_2.fastq.gz
## 17 ./data/SRR446043_1.fastq.gz ./data/SRR446043_2.fastq.gz
## 18 ./data/SRR446044_1.fastq.gz ./data/SRR446044_2.fastq.gz
##    SampleName Factor
## 1         M1A     M1
## 2         M1B     M1
## 3         A1A     A1
## 4         A1B     A1
## 5         V1A     V1
## 6         V1B     V1
## 7         M6A     M6
## 8         M6B     M6
## 9         A6A     A6
## 10        A6B     A6
## 11        V6A     V6
## 12        V6B     V6
## 13       M12A    M12
## 14       M12B    M12
## 15       A12A    A12
## 16       A12B    A12
## 17       V12A    V12
## 18       V12B    V12</code></pre>
</div>
</div>
<div id="read-preprocessing" class="section level1">
<h1 class="hasAnchor">
<a href="#read-preprocessing" class="anchor"></a>Read preprocessing</h1>
<div id="quality-filtering-and-adaptor-trimming" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-filtering-and-adaptor-trimming" class="anchor"></a>Quality filtering and adaptor trimming</h2>
<p>The following custom function trims adaptors hierarchically from the longest to the shortest match of the right end of the reads. If <code>internalmatch=TRUE</code> then internal matches will trigger the same behavior. The argument <code>minpatternlength</code> defines the shortest adaptor match to consider in this iterative process. In addition, the function removes reads containing Ns or homopolymer regions. More detailed information on read preprocessing is provided in <code>systemPipeR's</code> main vignette.</p>
<p>First, we construct <em><code>SYSargs2</code></em> object from <em><code>cwl</code></em> and <em><code>yml</code></em> param and <em><code>targets</code></em> files.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/preprocessReads/trim-pe"</span>, 
    package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetspath</span>, wf_file <span class="op">=</span> <span class="st">"trim-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"trim-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trim</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trim</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">output</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></pre></div>
<p>Next, we execute the code for trimming all the raw data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">fctpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"custom_Fct.R"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="va">fctpath</span><span class="op">)</span>
<span class="va">iterTrim</span> <span class="op">&lt;-</span> <span class="st">".iterTrimbatch1(fq, pattern='ACACGTCT', internalmatch=FALSE, minpatternlength=6, Nnumber=1, polyhomo=50, minreadlength=16, maxreadlength=101)"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">trim</span>, Fct <span class="op">=</span> <span class="va">iterTrim</span>, batchsize <span class="op">=</span> <span class="fl">1e+05</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span>, compress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">trim</span>, file <span class="op">=</span> <span class="st">"targets_trimPE.txt"</span>, step <span class="op">=</span> <span class="fl">1</span>, 
    new_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FileName1"</span>, <span class="st">"FileName2"</span><span class="op">)</span>, new_col_output_index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, 
        <span class="fl">2</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
</div>
<div id="fastq-quality-report" class="section level2">
<h2 class="hasAnchor">
<a href="#fastq-quality-report" class="anchor"></a>FASTQ quality report</h2>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of useful quality statistics for a set of FASTQ files including per cycle quality box plots, base proportions, base-level quality trends, relative k-mer diversity, length and occurrence distribution of reads, number of reads above quality cutoffs and mean quality distribution. The results are written to a PDF file named <code>fastqReport.png</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span>, batchsize <span class="op">=</span> <span class="fl">10000</span>, 
    klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"./results/fastqReport.png"</span>, height <span class="op">=</span> <span class="fl">18</span>, width <span class="op">=</span> <span class="fl">4</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span>, 
    units <span class="op">=</span> <span class="st">"in"</span>, res <span class="op">=</span> <span class="fl">72</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastqPlot</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<img src="results/fastqReport.png"><div align="center">
<b>Figure 1:</b> FASTQ quality report. To zoom in, right click image and open it in a separate browser tab.
</div>
<p></p>
</div>
</div>
<div id="alignments" class="section level1">
<h1 class="hasAnchor">
<a href="#alignments" class="anchor"></a>Alignments</h1>
<div id="read-mapping-with-hisat2" class="section level2">
<h2 class="hasAnchor">
<a href="#read-mapping-with-hisat2" class="anchor"></a>Read mapping with <code>HISAT2</code>
</h2>
<p>The following steps will demonstrate how to use the short read aligner <code>Hisat2</code> <span class="citation">(Kim, Langmead, and Salzberg 2015)</span> in both interactive job submissions and batch submissions to queuing systems of clusters using the <em><code>systemPipeR's</code></em> new CWL command-line interface.</p>
<p>Build <code>Hisat2</code> index.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/hisat2/hisat2-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"hisat2-index.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"hisat2-index.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>

<span class="co">## Run</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<p>The parameter settings of the aligner are defined in the <code>hisat2-mapping-se.cwl</code> and <code>hisat2-mapping-se.yml</code> files. The following shows how to construct the corresponding <em>SYSargs2</em> object, here <em>args</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/hisat2/hisat2-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetspath</span>, wf_file <span class="op">=</span> <span class="st">"hisat2-mapping-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"hisat2-mapping-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">args</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">output</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>

<span class="co">## Run</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">args</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">args</span>, 
    make_bam <span class="op">=</span> <span class="cn">TRUE</span>, dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, 
    template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, Njobs <span class="op">=</span> <span class="fl">18</span>, runid <span class="op">=</span> <span class="st">"01"</span>, 
    resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/waitForJobs.html">waitForJobs</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span></pre></div>
<p>Check whether all BAM files have been created.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span></pre></div>
</div>
<div id="read-and-alignment-stats" class="section level2">
<h2 class="hasAnchor">
<a href="#read-and-alignment-stats" class="anchor"></a>Read and alignment stats</h2>
<p>The following provides an overview of the number of reads in each sample and how many of them aligned to the reference.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/alignStats.html">alignStats</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">args</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">read_statsDF</span>, <span class="st">"results/alignStats.xls"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, 
    quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"alignStats.xls"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>, 
    header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="op">]</span></pre></div>
<pre><code>##   FileName Nreads2x Nalign Perc_Aligned Nalign_Primary
## 1      M1A   192918 177961     92.24697         177961
## 2      M1B   197484 159378     80.70426         159378
## 3      A1A   189870 176055     92.72397         176055
## 4      A1B   188854 147768     78.24457         147768
##   Perc_Aligned_Primary
## 1             92.24697
## 2             80.70426
## 3             92.72397
## 4             78.24457</code></pre>
</div>
<div id="create-symbolic-links-for-viewing-bam-files-in-igv" class="section level2">
<h2 class="hasAnchor">
<a href="#create-symbolic-links-for-viewing-bam-files-in-igv" class="anchor"></a>Create symbolic links for viewing BAM files in IGV</h2>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a genome browser such as IGV. The corresponding URLs are written to a file with a path specified under <code>urlfile</code> in the <code>results</code> directory.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/symLink2bam.html">symLink2bam</a></span><span class="op">(</span>sysargs <span class="op">=</span> <span class="va">args</span>, htmldir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"~/.html/"</span>, <span class="st">"projects/tests/"</span><span class="op">)</span>, 
    urlbase <span class="op">=</span> <span class="st">"http://biocluster.ucr.edu/~tgirke/"</span>, urlfile <span class="op">=</span> <span class="st">"./results/IGVurl.txt"</span><span class="op">)</span></pre></div>
</div>
</div>
<div id="read-distribution-across-genomic-features" class="section level1">
<h1 class="hasAnchor">
<a href="#read-distribution-across-genomic-features" class="anchor"></a>Read distribution across genomic features</h1>
<p>The <code>genFeatures</code> function generates a variety of feature types from <code>TxDb</code> objects using utilities provided by the <code>GenomicFeatures</code> package.</p>
<div id="obtain-feature-types" class="section level2">
<h2 class="hasAnchor">
<a href="#obtain-feature-types" class="anchor"></a>Obtain feature types</h2>
<p>The first step is the generation of the feature type ranges based on annotations provided by a GFF file that can be transformed into a <code>TxDb</code> object. This includes ranges for mRNAs, exons, introns, UTRs, CDSs, miRNAs, rRNAs, tRNAs, promoter and intergenic regions. In addition, any number of custom annotations can be included in this routine.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenomicFeatures</span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html">makeTxDbFromGFF</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/tair10.gff"</span>, format <span class="op">=</span> <span class="st">"gff3"</span>, 
    organism <span class="op">=</span> <span class="st">"Arabidopsis"</span><span class="op">)</span>
<span class="va">feat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/genFeatures.html">genFeatures</a></span><span class="op">(</span><span class="va">txdb</span>, featuretype <span class="op">=</span> <span class="st">"all"</span>, reduce_ranges <span class="op">=</span> <span class="cn">TRUE</span>, 
    upstream <span class="op">=</span> <span class="fl">1000</span>, downstream <span class="op">=</span> <span class="fl">0</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
</div>
<div id="count-and-plot-reads-of-any-length" class="section level2">
<h2 class="hasAnchor">
<a href="#count-and-plot-reads-of-any-length" class="anchor"></a>Count and plot reads of any length</h2>
<p>The <code>featuretypeCounts</code> function counts how many reads in short read alignment files (BAM format) overlap with entire annotation categories. This utility is useful for analyzing the distribution of the read mappings across feature types, <em>e.g.</em> coding versus non-coding genes. By default the read counts are reported for the sense and antisense strand of each feature type separately. To minimize memory consumption, the BAM files are processed in a stream using utilities from the <code>Rsamtools</code> and <code>GenomicAlignment</code> packages. The counts can be reported for each read length separately or as a single value for reads of any length. Subsequently, the counting results can be plotted with the associated <code>plotfeaturetypeCounts</code> function.</p>
<p>The following generates and plots feature counts for any read length.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">fc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/featuretypeCounts.html">featuretypeCounts</a></span><span class="op">(</span>bfl <span class="op">=</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span>, 
    grl <span class="op">=</span> <span class="va">feat</span>, singleEnd <span class="op">=</span> <span class="cn">TRUE</span>, readlength <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotfeaturetypeCounts.html">plotfeaturetypeCounts</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fc</span>, graphicsfile <span class="op">=</span> <span class="st">"results/featureCounts.png"</span>, 
    graphicsformat <span class="op">=</span> <span class="st">"png"</span>, scales <span class="op">=</span> <span class="st">"fixed"</span>, anyreadlength <span class="op">=</span> <span class="cn">TRUE</span>, 
    scale_length_val <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></pre></div>
<img src="results/featureCounts.png"><div align="center">
Figure 2: Read distribution plot across annotation features for any read length.
</div>
<p></p>
</div>
<div id="count-and-plot-reads-of-specific-lengths" class="section level2">
<h2 class="hasAnchor">
<a href="#count-and-plot-reads-of-specific-lengths" class="anchor"></a>Count and plot reads of specific lengths</h2>
<p>To determine the approximate read length of ribosome footprints in Ribo-Seq experiments, one can generate and plot the feature counts for specific read lengths separately. Typically, the most abundant read length obtained for translated features corresponds to the approximate footprint length occupied by the ribosomes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="va">fc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/featuretypeCounts.html">featuretypeCounts</a></span><span class="op">(</span>bfl <span class="op">=</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span>, 
    grl <span class="op">=</span> <span class="va">feat</span>, singleEnd <span class="op">=</span> <span class="cn">TRUE</span>, readlength <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">74</span><span class="op">:</span><span class="fl">76</span>, <span class="fl">99</span><span class="op">:</span><span class="fl">102</span><span class="op">)</span>, 
    type <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotfeaturetypeCounts.html">plotfeaturetypeCounts</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fc2</span>, graphicsfile <span class="op">=</span> <span class="st">"results/featureCounts2.png"</span>, 
    graphicsformat <span class="op">=</span> <span class="st">"png"</span>, scales <span class="op">=</span> <span class="st">"fixed"</span>, anyreadlength <span class="op">=</span> <span class="cn">FALSE</span>, 
    scale_length_val <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></pre></div>
<img src="results/featureCounts2.png"><div align="center">
Figure 3: Read distribution plot across annotation features for specific read lengths.
</div>
<p></p>
</div>
</div>
<div id="adding-custom-features-to-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#adding-custom-features-to-workflow" class="anchor"></a>Adding custom features to workflow</h1>
<div id="predicting-uorfs-in-5-utr-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#predicting-uorfs-in-5-utr-regions" class="anchor"></a>Predicting uORFs in 5’ UTR regions</h2>
<p>The function <code>predORF</code> can be used to identify open reading frames (ORFs) and coding sequences (CDSs) in DNA sequences provided as <code>DNAString</code> or <code>DNAStringSet</code> objects. The setting <code>mode='ORF'</code> returns continuous reading frames that begin with a start codon and end with a stop codon, while <code>mode='CDS'</code> returns continuous reading frames that do not need to begin or end with start or stop codons, respectively. Non-canonical start and stop condons are supported by allowing the user to provide any custom set of triplets under the <code>startcodon</code> and <code>stopcodon</code> arguments (<code>i.e.</code> non-ATG start codons). The argument <code>n</code> defines the maximum number of ORFs to return for each input sequence (<em>e.g.</em> <code>n=1</code> returns only the longest ORF). It also supports the identification of overlapping and nested ORFs. Alternatively, one can return all non-overlapping ORFs including the longest ORF for each input sequence with <code>n="all"</code> and <code>longest_disjoint=TRUE</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tgirke/systemPipeRdata">systemPipeRdata</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenomicFeatures</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html">makeTxDbFromGFF</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/tair10.gff"</span>, format <span class="op">=</span> <span class="st">"gff3"</span>, 
    organism <span class="op">=</span> <span class="st">"Arabidopsis"</span><span class="op">)</span>
<span class="va">futr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html">fiveUTRsByTranscript</a></span><span class="op">(</span><span class="va">txdb</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html">extractTranscriptSeqs</a></span><span class="op">(</span><span class="fu">FaFile</span><span class="op">(</span><span class="st">"data/tair10.fasta"</span><span class="op">)</span>, <span class="va">futr</span><span class="op">)</span>
<span class="va">uorf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/predORF.html">predORF</a></span><span class="op">(</span><span class="va">dna</span>, n <span class="op">=</span> <span class="st">"all"</span>, mode <span class="op">=</span> <span class="st">"orf"</span>, longest_disjoint <span class="op">=</span> <span class="cn">TRUE</span>, 
    strand <span class="op">=</span> <span class="st">"sense"</span><span class="op">)</span></pre></div>
<p>To use the predicted ORF ranges for expression analysis given genome alignments as input, it is necessary to scale them to the corresponding genome coordinates. The function <code>scaleRanges</code> does this by transforming the mappings of spliced features (query ranges) to their corresponding genome coordinates (subject ranges). The method accounts for introns in the subject ranges that are absent in the query ranges. The above uORFs predicted in the provided 5’ UTRs sequences using <code>predORF</code> are a typical use case for this application. These query ranges are given relative to the 5’ UTR sequences and <code>scaleRanges</code> will convert them to the corresponding genome coordinates. The resulting <code>GRangesList</code> object (here <code>grl_scaled</code>) can be directly used for read counting.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="va">grl_scaled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/scaleRanges.html">scaleRanges</a></span><span class="op">(</span>subject <span class="op">=</span> <span class="va">futr</span>, query <span class="op">=</span> <span class="va">uorf</span>, type <span class="op">=</span> <span class="st">"uORF"</span>, 
    verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html">export.gff3</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">grl_scaled</span><span class="op">)</span>, <span class="st">"results/uorf.gff"</span><span class="op">)</span></pre></div>
<p>To confirm the correctness of the obtained uORF ranges, one can parse their corresponding DNA sequences from the reference genome with the <code>getSeq</code> function and then translate them with the <code>translate</code> function into proteins. Typically, the returned protein sequences should start with a <code>M</code> (corresponding to start codon) and end with <code><a href="https://rdrr.io/r/base/Arithmetic.html">*</a></code> (corresponding to stop codon). The following example does this for a single uORF containing three exons.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="fu">translate</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu">getSeq</span><span class="op">(</span><span class="fu">FaFile</span><span class="op">(</span><span class="st">"data/tair10.fasta"</span><span class="op">)</span>, <span class="va">grl_scaled</span><span class="op">[[</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="adding-custom-features-to-other-feature-types" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-custom-features-to-other-feature-types" class="anchor"></a>Adding custom features to other feature types</h2>
<p>If required custom feature ranges can be added to the standard features generated with the <code>genFeatures</code> function above. The following does this for the uORF ranges predicted with <code>predORF</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="va">feat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/genFeatures.html">genFeatures</a></span><span class="op">(</span><span class="va">txdb</span>, featuretype <span class="op">=</span> <span class="st">"all"</span>, reduce_ranges <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">feat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">feat</span>, <span class="fu">GRangesList</span><span class="op">(</span>uORF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">grl_scaled</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="predicting-sorfs-in-intergenic-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#predicting-sorfs-in-intergenic-regions" class="anchor"></a>Predicting sORFs in intergenic regions</h2>
<p>The following identifies continuous ORFs in intergenic regions. Note, <code>predORF</code> can only identify continuous ORFs in query sequences. The function does not identify and remove introns prior to the ORF prediction.</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="va">feat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/genFeatures.html">genFeatures</a></span><span class="op">(</span><span class="va">txdb</span>, featuretype <span class="op">=</span> <span class="st">"intergenic"</span>, reduce_ranges <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">intergenic</span> <span class="op">&lt;-</span> <span class="va">feat</span><span class="op">$</span><span class="va">intergenic</span>
<span class="fu">strand</span><span class="op">(</span><span class="va">intergenic</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"+"</span>
<span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu">getSeq</span><span class="op">(</span><span class="fu">FaFile</span><span class="op">(</span><span class="st">"data/tair10.fasta"</span><span class="op">)</span>, <span class="va">intergenic</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">intergenic</span><span class="op">)</span><span class="op">$</span><span class="va">feature_by</span>
<span class="va">sorf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/predORF.html">predORF</a></span><span class="op">(</span><span class="va">dna</span>, n <span class="op">=</span> <span class="st">"all"</span>, mode <span class="op">=</span> <span class="st">"orf"</span>, longest_disjoint <span class="op">=</span> <span class="cn">TRUE</span>, 
    strand <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span>
<span class="va">sorf</span> <span class="op">&lt;-</span> <span class="va">sorf</span><span class="op">[</span><span class="fu">width</span><span class="op">(</span><span class="va">sorf</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">60</span><span class="op">]</span>  <span class="co"># Remove sORFs below length cutoff, here 60bp</span>
<span class="va">intergenic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">intergenic</span>, <span class="fu">mcols</span><span class="op">(</span><span class="va">intergenic</span><span class="op">)</span><span class="op">$</span><span class="va">feature_by</span><span class="op">)</span>
<span class="va">grl_scaled_intergenic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/scaleRanges.html">scaleRanges</a></span><span class="op">(</span>subject <span class="op">=</span> <span class="va">intergenic</span>, query <span class="op">=</span> <span class="va">sorf</span>, 
    type <span class="op">=</span> <span class="st">"sORF"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html">export.gff3</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">grl_scaled_intergenic</span><span class="op">)</span>, <span class="st">"sorf.gff"</span><span class="op">)</span>
<span class="fu">translate</span><span class="op">(</span><span class="fu">getSeq</span><span class="op">(</span><span class="fu">FaFile</span><span class="op">(</span><span class="st">"data/tair10.fasta"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">grl_scaled_intergenic</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
</div>
<div id="genomic-read-coverage-along-transripts-or-cdss" class="section level1">
<h1 class="hasAnchor">
<a href="#genomic-read-coverage-along-transripts-or-cdss" class="anchor"></a>Genomic read coverage along transripts or CDSs</h1>
<p>The <code>featureCoverage</code> function computes the read coverage along single and multi component features based on genomic alignments. The coverage segments of component features are spliced to continuous ranges, such as exons to transcripts or CDSs to ORFs. The results can be obtained with single nucleotide resolution (<em>e.g.</em> around start and stop codons) or as mean coverage of relative bin sizes, such as 100 bins for each feature. The latter allows comparisons of coverage trends among transcripts of variable length. Additionally, the results can be obtained for single or many features (<em>e.g.</em> any number of transcripts) at once. Visualization of the coverage results is facilitated by the downstream <code>plotfeatureCoverage</code> function.</p>
<div id="binned-cds-coverage-to-compare-many-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#binned-cds-coverage-to-compare-many-transcripts" class="anchor"></a>Binned CDS coverage to compare many transcripts</h2>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="va">grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html">cdsBy</a></span><span class="op">(</span><span class="va">txdb</span>, <span class="st">"tx"</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">fcov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/featureCoverage.html">featureCoverage</a></span><span class="op">(</span>bfl <span class="op">=</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, grl <span class="op">=</span> <span class="va">grl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, 
    resizereads <span class="op">=</span> <span class="cn">NULL</span>, readlengthrange <span class="op">=</span> <span class="cn">NULL</span>, Nbins <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="va">mean</span>, 
    fixedmatrix <span class="op">=</span> <span class="cn">FALSE</span>, resizefeatures <span class="op">=</span> <span class="cn">TRUE</span>, upstream <span class="op">=</span> <span class="fl">20</span>, 
    downstream <span class="op">=</span> <span class="fl">20</span>, outfile <span class="op">=</span> <span class="st">"results/featureCoverage.xls"</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
</div>
<div id="coverage-upstream-and-downstream-of-start-and-stop-codons" class="section level2">
<h2 class="hasAnchor">
<a href="#coverage-upstream-and-downstream-of-start-and-stop-codons" class="anchor"></a>Coverage upstream and downstream of start and stop codons</h2>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="va">fcov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/featureCoverage.html">featureCoverage</a></span><span class="op">(</span>bfl <span class="op">=</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>, grl <span class="op">=</span> <span class="va">grl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>, 
    resizereads <span class="op">=</span> <span class="cn">NULL</span>, readlengthrange <span class="op">=</span> <span class="cn">NULL</span>, Nbins <span class="op">=</span> <span class="cn">NULL</span>, 
    method <span class="op">=</span> <span class="va">mean</span>, fixedmatrix <span class="op">=</span> <span class="cn">TRUE</span>, resizefeatures <span class="op">=</span> <span class="cn">TRUE</span>, 
    upstream <span class="op">=</span> <span class="fl">20</span>, downstream <span class="op">=</span> <span class="fl">20</span>, outfile <span class="op">=</span> <span class="st">"results/featureCoverage.xls"</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotfeatureCoverage.html">plotfeatureCoverage</a></span><span class="op">(</span>covMA <span class="op">=</span> <span class="va">fcov</span>, method <span class="op">=</span> <span class="va">mean</span>, scales <span class="op">=</span> <span class="st">"fixed"</span>, 
    extendylim <span class="op">=</span> <span class="fl">2</span>, scale_count_val <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="fl">6</span><span class="op">)</span></pre></div>
</div>
<div id="combined-coverage-for-both-binned-cds-and-startstop-codons" class="section level2">
<h2 class="hasAnchor">
<a href="#combined-coverage-for-both-binned-cds-and-startstop-codons" class="anchor"></a>Combined coverage for both binned CDS and start/stop codons</h2>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span>
<span class="va">fcov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/featureCoverage.html">featureCoverage</a></span><span class="op">(</span>bfl <span class="op">=</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>, grl <span class="op">=</span> <span class="va">grl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, 
    resizereads <span class="op">=</span> <span class="cn">NULL</span>, readlengthrange <span class="op">=</span> <span class="cn">NULL</span>, Nbins <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="va">mean</span>, 
    fixedmatrix <span class="op">=</span> <span class="cn">TRUE</span>, resizefeatures <span class="op">=</span> <span class="cn">TRUE</span>, upstream <span class="op">=</span> <span class="fl">20</span>, 
    downstream <span class="op">=</span> <span class="fl">20</span>, outfile <span class="op">=</span> <span class="st">"results/featureCoverage.xls"</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"./results/featurePlot.png"</span>, height <span class="op">=</span> <span class="fl">12</span>, width <span class="op">=</span> <span class="fl">24</span>, units <span class="op">=</span> <span class="st">"in"</span>, 
    res <span class="op">=</span> <span class="fl">72</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotfeatureCoverage.html">plotfeatureCoverage</a></span><span class="op">(</span>covMA <span class="op">=</span> <span class="va">fcov</span>, method <span class="op">=</span> <span class="va">mean</span>, scales <span class="op">=</span> <span class="st">"fixed"</span>, 
    extendylim <span class="op">=</span> <span class="fl">2</span>, scale_count_val <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="fl">6</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<img src="results/featurePlot.png"><div align="center">
Figure 4: Feature coverage plot with single nucleotide resolution around start and stop codons and binned coverage between them.
</div>
<p></p>
</div>
<div id="nucleotide-level-coverage-along-entire-transcriptscdss" class="section level2">
<h2 class="hasAnchor">
<a href="#nucleotide-level-coverage-along-entire-transcriptscdss" class="anchor"></a>Nucleotide level coverage along entire transcripts/CDSs</h2>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="va">fcov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/featureCoverage.html">featureCoverage</a></span><span class="op">(</span>bfl <span class="op">=</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, grl <span class="op">=</span> <span class="va">grl</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, 
    resizereads <span class="op">=</span> <span class="cn">NULL</span>, readlengthrange <span class="op">=</span> <span class="cn">NULL</span>, Nbins <span class="op">=</span> <span class="cn">NULL</span>, 
    method <span class="op">=</span> <span class="va">mean</span>, fixedmatrix <span class="op">=</span> <span class="cn">FALSE</span>, resizefeatures <span class="op">=</span> <span class="cn">TRUE</span>, 
    upstream <span class="op">=</span> <span class="fl">20</span>, downstream <span class="op">=</span> <span class="fl">20</span>, outfile <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></pre></div>
</div>
</div>
<div id="read-quantification-per-annotation-range" class="section level1">
<h1 class="hasAnchor">
<a href="#read-quantification-per-annotation-range" class="anchor"></a>Read quantification per annotation range</h1>
<div id="read-counting-with-summarizeoverlaps-in-parallel-mode-using-multiple-cores" class="section level2">
<h2 class="hasAnchor">
<a href="#read-counting-with-summarizeoverlaps-in-parallel-mode-using-multiple-cores" class="anchor"></a>Read counting with <code>summarizeOverlaps</code> in parallel mode using multiple cores</h2>
<p>Reads overlapping with annotation ranges of interest are counted for each sample using the <code>summarizeOverlaps</code> function <span class="citation">(Lawrence et al. 2013)</span>. The read counting is preformed for exonic gene regions in a non-strand-specific manner while ignoring overlaps among different genes. Subsequently, the expression count values are normalized by  (RPKM). The raw read count table (<code>countDFeByg.xls</code>) and the corresponding RPKM table (<code>rpkmDFeByg.xls</code>) are written to separate files in the <code>results</code> directory of this project. Parallelization is achieved with the <code>BiocParallel</code> package, here using 8 CPU cores.</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"GenomicFeatures"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">loadDb</span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span>
<span class="va">eByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html">exonsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span><span class="op">)</span>
<span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu">BamFileList</span><span class="op">(</span><span class="va">outpaths</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">multicoreParam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span><span class="op">(</span><span class="va">multicoreParam</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">registered</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">counteByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="va">bfl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">summarizeOverlaps</span><span class="op">(</span><span class="va">eByg</span>, 
    <span class="va">x</span>, mode <span class="op">=</span> <span class="st">"Union"</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span>, inter.feature <span class="op">=</span> <span class="cn">FALSE</span>, 
    singleEnd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="va">countDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">counteByg</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">assays</span><span class="op">(</span><span class="va">counteByg</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">countDFeByg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">counteByg</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">countDFeByg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span>
<span class="va">rpkmDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">countDFeByg</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/returnRPKM.html">returnRPKM</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">x</span>, 
    ranges <span class="op">=</span> <span class="va">eByg</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">countDFeByg</span>, <span class="st">"results/countDFeByg.xls"</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, 
    quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">rpkmDFeByg</span>, <span class="st">"results/rpkmDFeByg.xls"</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, 
    quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></pre></div>
<p>Sample of data slice of count table</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"results/countDFeByg.xls"</span>, row.names <span class="op">=</span> <span class="fl">1</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, 
    <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></pre></div>
<p>Sample of data slice of RPKM table</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"results/rpkmDFeByg.xls"</span>, row.names <span class="op">=</span> <span class="fl">1</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, 
    <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></pre></div>
<p>Note, for most statistical differential expression or abundance analysis methods, such as <code>edgeR</code> or <code>DESeq2</code>, the raw count values should be used as input. The usage of RPKM values should be restricted to specialty applications required by some users, <em>e.g.</em> manually comparing the expression levels among different genes or features.</p>
</div>
<div id="sample-wise-correlation-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-wise-correlation-analysis" class="anchor"></a>Sample-wise correlation analysis</h2>
<p>The following computes the sample-wise Spearman correlation coefficients from the <code>rlog</code> transformed expression values generated with the <code>DESeq2</code> package. After transformation to a distance matrix, hierarchical clustering is performed with the <code>hclust</code> function and the result is plotted as a dendrogram and written to a file named <code>sample_tree.png</code> in the <code>results</code> directory.</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/">ape</a></span>, warn.conflicts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">countDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"./results/countDFeByg.xls"</span><span class="op">)</span><span class="op">)</span>
<span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/targets.html">targets</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">SampleName</span>, 
    condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/targets.html">targets</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">Factor</span><span class="op">)</span>
<span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">countDF</span>, colData <span class="op">=</span> <span class="va">colData</span>, 
    design <span class="op">=</span> <span class="op">~</span><span class="va">condition</span><span class="op">)</span>
<span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/rlog.html">rlog</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span>
<span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">d</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"results/sample_tree.pdf"</span><span class="op">)</span>
<span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html">plot.phylo</a></span><span class="op">(</span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html">as.phylo</a></span><span class="op">(</span><span class="va">hc</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"p"</span>, edge.col <span class="op">=</span> <span class="st">"blue"</span>, 
    edge.width <span class="op">=</span> <span class="fl">2</span>, show.node.label <span class="op">=</span> <span class="cn">TRUE</span>, no.margin <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<img src="results/sample_tree.png"><div align="center">
Figure 5: Correlation dendrogram of samples.
</div>
<p></p>
</div>
</div>
<div id="analysis-of-differentially-expressed-genes-with-edger" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis-of-differentially-expressed-genes-with-edger" class="anchor"></a>Analysis of differentially expressed genes with <code>edgeR</code>
</h1>
<p>The analysis of differentially expressed genes (DEGs) is performed with the glm method from the <code>edgeR</code> package <span class="citation">(Robinson, McCarthy, and Smyth 2010)</span>. The sample comparisons used by this analysis are defined in the header lines of the <code>targetsPE.txt</code> file starting with <code>&lt;CMP&gt;</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">countDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"results/countDFeByg.xls"</span>, row.names <span class="op">=</span> <span class="fl">1</span>, 
    check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"targetsPE.txt"</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>
<span class="va">cmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/readComp.html">readComp</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"targetsPE.txt"</span>, format <span class="op">=</span> <span class="st">"matrix"</span>, delim <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span>
<span class="va">edgeDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/run_edgeR.html">run_edgeR</a></span><span class="op">(</span>countDF <span class="op">=</span> <span class="va">countDF</span>, targets <span class="op">=</span> <span class="va">targets</span>, cmp <span class="op">=</span> <span class="va">cmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, 
    independent <span class="op">=</span> <span class="cn">FALSE</span>, mdsplot <span class="op">=</span> <span class="st">""</span><span class="op">)</span></pre></div>
<p>Add functional gene descriptions, here from <code>biomaRt</code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"plants_mart"</span>, dataset <span class="op">=</span> <span class="st">"athaliana_eg_gene"</span>, host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"tair_locus"</span>, <span class="st">"description"</span><span class="op">)</span>, mart <span class="op">=</span> <span class="va">m</span><span class="op">)</span>
<span class="va">desc</span> <span class="op">&lt;-</span> <span class="va">desc</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">desc</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">descv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">desc</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">descv</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">desc</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">edgeDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">edgeDF</span>, Desc <span class="op">=</span> <span class="va">descv</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">edgeDF</span><span class="op">)</span><span class="op">]</span>, 
    check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">edgeDF</span>, <span class="st">"./results/edgeRglm_allcomp.xls"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
    sep <span class="op">=</span> <span class="st">"\t"</span>, col.names <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></pre></div>
<p>Filter and plot DEG results for up and down regulated genes. The definition of <code>up</code> and <code>down</code> is given in the corresponding help file. To open it, type <code><a href="https://rdrr.io/pkg/systemPipeR/man/filterDEGs.html">?filterDEGs</a></code> in the R console.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="va">edgeDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"results/edgeRglm_allcomp.xls"</span>, row.names <span class="op">=</span> <span class="fl">1</span>, 
    check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"./results/DEGcounts.png"</span>, height <span class="op">=</span> <span class="fl">10</span>, width <span class="op">=</span> <span class="fl">10</span>, units <span class="op">=</span> <span class="st">"in"</span>, 
    res <span class="op">=</span> <span class="fl">72</span><span class="op">)</span>
<span class="va">DEG_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/filterDEGs.html">filterDEGs</a></span><span class="op">(</span>degDF <span class="op">=</span> <span class="va">edgeDF</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Fold <span class="op">=</span> <span class="fl">2</span>, FDR <span class="op">=</span> <span class="fl">20</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">$</span><span class="va">Summary</span>, <span class="st">"./results/DEGcounts.xls"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
    sep <span class="op">=</span> <span class="st">"\t"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<img src="results/DEGcounts.png"><div align="center">
Figure 6: Up and down regulated DEGs.
</div>
<p></p>
<p>The function <code>overLapper</code> can compute Venn intersects for large numbers of sample sets (up to 20 or more) and <code>vennPlot</code> can plot 2-5 way Venn diagrams. A useful feature is the possiblity to combine the counts from several Venn comparisons with the same number of sample sets in a single Venn diagram (here for 4 up and down DEG sets).</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="va">vennsetup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/overLapper.html">overLapper</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">$</span><span class="va">Up</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"vennsets"</span><span class="op">)</span>
<span class="va">vennsetdown</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/overLapper.html">overLapper</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">$</span><span class="va">Down</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"vennsets"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"results/vennplot.png"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/vennPlot.html">vennPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">vennsetup</span>, <span class="va">vennsetdown</span><span class="op">)</span>, mymain <span class="op">=</span> <span class="st">""</span>, mysub <span class="op">=</span> <span class="st">""</span>, 
    colmode <span class="op">=</span> <span class="fl">2</span>, ccol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<img src="results/vennplot.png"><div align="center">
Figure 7: Venn Diagram for 4 Up and Down DEG Sets
</div>
<p></p>
</div>
<div id="go-term-enrichment-analysis-of-degs" class="section level1">
<h1 class="hasAnchor">
<a href="#go-term-enrichment-analysis-of-degs" class="anchor"></a>GO term enrichment analysis of DEGs</h1>
<div id="obtain-gene-to-go-mappings" class="section level2">
<h2 class="hasAnchor">
<a href="#obtain-gene-to-go-mappings" class="anchor"></a>Obtain gene-to-GO mappings</h2>
<p>The following shows how to obtain gene-to-GO mappings from <code>biomaRt</code> (here for <em>A. thaliana</em>) and how to organize them for the downstream GO term enrichment analysis. Alternatively, the gene-to-GO mappings can be obtained for many organisms from Bioconductor’s <code>*.db</code> genome annotation packages or GO annotation files provided by various genome databases. For each annotation this relatively slow preprocessing step needs to be performed only once. Subsequently, the preprocessed data can be loaded with the <code>load</code> function as shown in the next subsection.</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listMarts.html">listMarts</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># To choose BioMart database</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listMarts.html">listMarts</a></span><span class="op">(</span>host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"plants_mart"</span>, host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html">listDatasets</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"plants_mart"</span>, dataset <span class="op">=</span> <span class="st">"athaliana_eg_gene"</span>, host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listAttributes.html">listAttributes</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>  <span class="co"># Choose data types you want to download</span>
<span class="va">go</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"go_id"</span>, <span class="st">"tair_locus"</span>, <span class="st">"namespace_1003"</span><span class="op">)</span>, 
    mart <span class="op">=</span> <span class="va">m</span><span class="op">)</span>
<span class="va">go</span> <span class="op">&lt;-</span> <span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">!=</span> <span class="st">""</span>, <span class="op">]</span>
<span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span>
<span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">==</span> <span class="st">"molecular_function"</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"F"</span>
<span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">==</span> <span class="st">"biological_process"</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P"</span>
<span class="va">go</span><span class="op">[</span><span class="va">go</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">==</span> <span class="st">"cellular_component"</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"C"</span>
<span class="va">go</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="st">"./data/GO"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">go</span>, <span class="st">"data/GO/GOannotationsBiomart_mod.txt"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
    row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="va">catdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html">makeCATdb</a></span><span class="op">(</span>myfile <span class="op">=</span> <span class="st">"data/GO/GOannotationsBiomart_mod.txt"</span>, 
    lib <span class="op">=</span> <span class="cn">NULL</span>, org <span class="op">=</span> <span class="st">""</span>, colno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, idconv <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">catdb</span>, file <span class="op">=</span> <span class="st">"data/GO/catdb.RData"</span><span class="op">)</span></pre></div>
</div>
<div id="batch-go-term-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#batch-go-term-enrichment-analysis" class="anchor"></a>Batch GO term enrichment analysis</h2>
<p>Apply the enrichment analysis to the DEG sets obtained the above differential expression analysis. Note, in the following example the <code>FDR</code> filter is set here to an unreasonably high value, simply because of the small size of the toy data set used in this vignette. Batch enrichment analysis of many gene sets is performed with the function. When <code>method=all</code>, it returns all GO terms passing the p-value cutoff specified under the <code>cutoff</code> arguments. When <code>method=slim</code>, it returns only the GO terms specified under the <code>myslimv</code> argument. The given example shows how a GO slim vector for a specific organism can be obtained from BioMart.</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/berndbischl/BBmisc">BBmisc</a></span><span class="op">)</span>  <span class="co"># Defines suppressAll()</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/GO/catdb.RData"</span><span class="op">)</span>
<span class="va">DEG_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/filterDEGs.html">filterDEGs</a></span><span class="op">(</span>degDF <span class="op">=</span> <span class="va">edgeDF</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Fold <span class="op">=</span> <span class="fl">2</span>, FDR <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, 
    plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">up_down</span> <span class="op">&lt;-</span> <span class="va">DEG_list</span><span class="op">$</span><span class="va">UporDown</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up_down</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up_down</span><span class="op">)</span>, <span class="st">"_up_down"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="va">up</span> <span class="op">&lt;-</span> <span class="va">DEG_list</span><span class="op">$</span><span class="va">Up</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">up</span><span class="op">)</span>, <span class="st">"_up"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="va">down</span> <span class="op">&lt;-</span> <span class="va">DEG_list</span><span class="op">$</span><span class="va">Down</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">down</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">down</span><span class="op">)</span>, <span class="st">"_down"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="va">DEGlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">up_down</span>, <span class="va">up</span>, <span class="va">down</span><span class="op">)</span>
<span class="va">DEGlist</span> <span class="op">&lt;-</span> <span class="va">DEGlist</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">DEGlist</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span>
<span class="va">BatchResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html">GOCluster_Report</a></span><span class="op">(</span>catdb <span class="op">=</span> <span class="va">catdb</span>, setlist <span class="op">=</span> <span class="va">DEGlist</span>, 
    method <span class="op">=</span> <span class="st">"all"</span>, id_type <span class="op">=</span> <span class="st">"gene"</span>, CLSZ <span class="op">=</span> <span class="fl">2</span>, cutoff <span class="op">=</span> <span class="fl">0.9</span>, 
    gocats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MF"</span>, <span class="st">"BP"</span>, <span class="st">"CC"</span><span class="op">)</span>, recordSpecGO <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"plants_mart"</span>, dataset <span class="op">=</span> <span class="st">"athaliana_eg_gene"</span>, host <span class="op">=</span> <span class="st">"plants.ensembl.org"</span><span class="op">)</span>
<span class="va">goslimvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"goslim_goa_accession"</span><span class="op">)</span>, 
    mart <span class="op">=</span> <span class="va">m</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">BatchResultslim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html">GOCluster_Report</a></span><span class="op">(</span>catdb <span class="op">=</span> <span class="va">catdb</span>, setlist <span class="op">=</span> <span class="va">DEGlist</span>, 
    method <span class="op">=</span> <span class="st">"slim"</span>, id_type <span class="op">=</span> <span class="st">"gene"</span>, myslimv <span class="op">=</span> <span class="va">goslimvec</span>, CLSZ <span class="op">=</span> <span class="fl">10</span>, 
    cutoff <span class="op">=</span> <span class="fl">0.01</span>, gocats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MF"</span>, <span class="st">"BP"</span>, <span class="st">"CC"</span><span class="op">)</span>, recordSpecGO <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></pre></div>
<p>The <code>data.frame</code> generated by <code>GOCluster</code> can be plotted with the <code>goBarplot</code> function. Because of the variable size of the sample sets, it may not always be desirable to show the results from different DEG sets in the same bar plot. Plotting single sample sets is achieved by subsetting the input data frame as shown in the first line of the following example.</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="va">gos</span> <span class="op">&lt;-</span> <span class="va">BatchResultslim</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"M6-V6_up_down"</span>, <span class="va">BatchResultslim</span><span class="op">$</span><span class="va">CLID</span><span class="op">)</span>, 
    <span class="op">]</span>
<span class="va">gos</span> <span class="op">&lt;-</span> <span class="va">BatchResultslim</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"./results/GOslimbarplotMF.png"</span>, height <span class="op">=</span> <span class="fl">12</span>, width <span class="op">=</span> <span class="fl">12</span>, 
    units <span class="op">=</span> <span class="st">"in"</span>, res <span class="op">=</span> <span class="fl">72</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html">goBarplot</a></span><span class="op">(</span><span class="va">gos</span>, gocat <span class="op">=</span> <span class="st">"MF"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html">goBarplot</a></span><span class="op">(</span><span class="va">gos</span>, gocat <span class="op">=</span> <span class="st">"BP"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html">goBarplot</a></span><span class="op">(</span><span class="va">gos</span>, gocat <span class="op">=</span> <span class="st">"CC"</span><span class="op">)</span></pre></div>
<img src="results/GOslimbarplotMF.png"><div align="center">
Figure 8: GO Slim Barplot for MF Ontology.
</div>
<p></p>
</div>
</div>
<div id="differential-ribosome-loading-analysis-translational-efficiency" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-ribosome-loading-analysis-translational-efficiency" class="anchor"></a>Differential ribosome loading analysis (translational efficiency)</h1>
<p>Combined with mRNA-Seq data, Ribo-Seq or polyRibo-Seq experiments can be used to study changes in translational efficiencies of genes and/or transcripts for different treatments. For test purposes the following generates a small test data set from the sample data used in this vignette, where two types of RNA samples (<code>assays</code>) are considered: polyribosomal mRNA (<code>Ribo</code>) and total mRNA (<code>mRNA</code>). In addition, there are two treatments (<code>conditions</code>): <code>M1</code> and <code>A1</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span>
<span class="va">countDFeBygpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"countDFeByg.xls"</span>, 
    package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">countDFeByg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">countDFeBygpath</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span>assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Ribo"</span>, <span class="st">"mRNA"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>, 
    condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/targets.html">targets</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">Factor</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>, 
        <span class="fl">2</span><span class="op">)</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/targets.html">targets</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">SampleName</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span>
<span class="va">coldata</span></pre></div>
<p>Differences in translational efficiencies can be calculated by ratios of ratios for the two conditions:</p>
<p><span class="math display">\[(Ribo\_A1 / mRNA\_A1) / (Ribo\_M1 / mRNA\_M1)\]</span></p>
<p>The latter can be modeled with the <code>DESeq2</code> package using the design <span class="math inline">\(\sim assay + condition + assay:condition\)</span>, where the interaction term <span class="math inline">\(assay:condition\)</span> represents the ratio of ratios. Using the likelihood ratio test of <code>DESeq2</code>, which removes the interaction term in the reduced model, one can test whether the translational efficiency (ribosome loading) is different in condition <code>A1</code> than in <code>M1</code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">countDFeByg</span><span class="op">[</span>, 
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, colData <span class="op">=</span> <span class="va">coldata</span>, design <span class="op">=</span> <span class="op">~</span><span class="va">assay</span> <span class="op">+</span> 
    <span class="va">condition</span> <span class="op">+</span> <span class="va">assay</span><span class="op">:</span><span class="va">condition</span><span class="op">)</span>
<span class="co"># model.matrix(~ assay + condition + assay:condition,</span>
<span class="co"># coldata) # Corresponding design matrix</span>
<span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span><span class="op">(</span><span class="va">dds</span>, test <span class="op">=</span> <span class="st">"LRT"</span>, reduced <span class="op">=</span> <span class="op">~</span><span class="va">assay</span> <span class="op">+</span> <span class="va">condition</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span>, <span class="op">]</span>, <span class="fl">4</span><span class="op">)</span>
<span class="co"># write.table(res, file='transleff.xls', quote=FALSE,</span>
<span class="co"># col.names = NA, sep='\t')</span></pre></div>
</div>
<div id="clustering-and-heat-maps" class="section level1">
<h1 class="hasAnchor">
<a href="#clustering-and-heat-maps" class="anchor"></a>Clustering and heat maps</h1>
<p>The following example performs hierarchical clustering on the <code>rlog</code> transformed expression matrix subsetted by the DEGs identified in the above differential expression analysis. It uses a Pearson correlation-based distance measure and complete linkage for cluster joining.</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>
<span class="va">geneids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">DEG_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/rlog.html">rlog</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">geneids</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"heatmap1.png"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">y</span>, scale <span class="op">=</span> <span class="st">"row"</span>, clustering_distance_rows <span class="op">=</span> <span class="st">"correlation"</span>, 
    clustering_distance_cols <span class="op">=</span> <span class="st">"correlation"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<img src="results/heatmap1.png"><div align="center">
Figure 9: Heat map with hierarchical clustering dendrograms of DEGs
</div>
<p></p>
</div>
<div id="render-report-in-html-and-pdf-format" class="section level1">
<h1 class="hasAnchor">
<a href="#render-report-in-html-and-pdf-format" class="anchor"></a>Render report in HTML and PDF format</h1>
<div class="sourceCode" id="cb41"><pre class="downlit">
<span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rmarkdown/man/render.html">render</a></span><span class="op">(</span><span class="st">"systemPipeRIBOseq.Rmd"</span>, <span class="st">"html_document"</span><span class="op">)</span>
<span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rmarkdown/man/render.html">render</a></span><span class="op">(</span><span class="st">"systemPipeRIBOseq.Rmd"</span>, <span class="st">"pdf_document"</span><span class="op">)</span></pre></div>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version Information</h1>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>## R Under development (unstable) (2020-08-07 r78979)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-devel/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices
## [6] utils     datasets  methods   base     
## 
## other attached packages:
##  [1] systemPipeR_1.23.4          ShortRead_1.47.2           
##  [3] GenomicAlignments_1.25.3    SummarizedExperiment_1.19.9
##  [5] Biobase_2.49.1              MatrixGenerics_1.1.5       
##  [7] matrixStats_0.57.0          BiocParallel_1.23.2        
##  [9] Rsamtools_2.5.3             Biostrings_2.57.2          
## [11] XVector_0.29.3              GenomicRanges_1.41.6       
## [13] GenomeInfoDb_1.25.11        IRanges_2.23.10            
## [15] S4Vectors_0.27.14           BiocGenerics_0.35.4        
## [17] BiocStyle_2.17.1           
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_1.4-1         rjson_0.2.20            
##   [3] hwriter_1.3.2            ellipsis_0.3.1          
##   [5] rprojroot_1.3-2          fs_1.5.0                
##   [7] rstudioapi_0.11          bit64_4.0.5             
##   [9] AnnotationDbi_1.51.3     xml2_1.3.2              
##  [11] codetools_0.2-16         splines_4.1.0           
##  [13] knitr_1.30               jsonlite_1.7.1          
##  [15] annotate_1.67.1          GO.db_3.11.4            
##  [17] dbplyr_1.4.4             png_0.1-7               
##  [19] pheatmap_1.0.12          graph_1.67.1            
##  [21] BiocManager_1.30.10      compiler_4.1.0          
##  [23] httr_1.4.2               GOstats_2.55.0          
##  [25] backports_1.1.10         assertthat_0.2.1        
##  [27] Matrix_1.2-18            limma_3.45.14           
##  [29] formatR_1.7              htmltools_0.5.0         
##  [31] prettyunits_1.1.1        tools_4.1.0             
##  [33] gtable_0.3.0             glue_1.4.2              
##  [35] GenomeInfoDbData_1.2.4   Category_2.55.0         
##  [37] dplyr_1.0.2              rsvg_2.1                
##  [39] batchtools_0.9.13        rappdirs_0.3.1          
##  [41] V8_3.2.0                 Rcpp_1.0.5              
##  [43] pkgdown_1.6.1            vctrs_0.3.4             
##  [45] rtracklayer_1.49.5       xfun_0.18               
##  [47] stringr_1.4.0            lifecycle_0.2.0         
##  [49] XML_3.99-0.5             edgeR_3.31.4            
##  [51] zlibbioc_1.35.0          scales_1.1.1            
##  [53] BSgenome_1.57.7          VariantAnnotation_1.35.3
##  [55] ragg_0.4.0               hms_0.5.3               
##  [57] RBGL_1.65.0              RColorBrewer_1.1-2      
##  [59] yaml_2.2.1               curl_4.3                
##  [61] memoise_1.1.0            ggplot2_3.3.2           
##  [63] biomaRt_2.45.7           latticeExtra_0.6-29     
##  [65] stringi_1.5.3            RSQLite_2.2.1           
##  [67] genefilter_1.71.0        desc_1.2.0              
##  [69] checkmate_2.0.0          GenomicFeatures_1.41.3  
##  [71] DOT_0.1                  rlang_0.4.8             
##  [73] pkgconfig_2.0.3          systemfonts_0.3.2       
##  [75] bitops_1.0-6             evaluate_0.14           
##  [77] lattice_0.20-41          purrr_0.3.4             
##  [79] bit_4.0.4                tidyselect_1.1.0        
##  [81] GSEABase_1.51.1          AnnotationForge_1.31.3  
##  [83] magrittr_1.5             bookdown_0.21           
##  [85] R6_2.4.1                 generics_0.0.2          
##  [87] base64url_1.4            DelayedArray_0.15.16    
##  [89] DBI_1.1.0                withr_2.3.0             
##  [91] pillar_1.4.6             survival_3.2-7          
##  [93] RCurl_1.98-1.2           tibble_3.0.4            
##  [95] crayon_1.3.4             BiocFileCache_1.13.1    
##  [97] rmarkdown_2.4            jpeg_0.1-8.1            
##  [99] progress_1.2.2           locfit_1.5-9.4          
## [101] grid_4.1.0               data.table_1.13.0       
## [103] Rgraphviz_2.33.0         blob_1.2.1              
## [105] digest_0.6.25            xtable_1.8-4            
## [107] brew_1.0-6               textshaping_0.1.2       
## [109] openssl_1.4.3            munsell_0.5.0           
## [111] askpass_1.1</code></pre>
</div>
<div id="funding" class="section level1">
<h1 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h1>
<p>This research was funded by National Science Foundation Grants IOS-0750811 and MCB-1021969, and a Marie Curie European Economic Community Fellowship PIOF-GA-2012-327954.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-Aspden2014-uu">
<p>Aspden, Julie L, Ying Chen Eyre-Walker, Rose J Phillips, Unum Amin, Muhammad Ali S Mumtaz, Michele Brocard, and Juan-Pablo Couso. 2014. “Extensive Translation of Small Open Reading Frames Revealed by Poly-Ribo-Seq.” <em>Elife</em> 3: e03528. <a href="https://doi.org/10.7554/eLife.03528">https://doi.org/10.7554/eLife.03528</a>.</p>
</div>
<div id="ref-H_Backman2016-bt">
<p>H Backman, Tyler W, and Thomas Girke. 2016. “systemPipeR: NGS workflow and report generation environment.” <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.</p>
</div>
<div id="ref-Ingolia2009-cb">
<p>Ingolia, N T, S Ghaemmaghami, J R Newman, and J S Weissman. 2009. “Genome-Wide Analysis in Vivo of Translation with Nucleotide Resolution Using Ribosome Profiling.” <em>Science</em> 324 (5924): 218–23. <a href="https://doi.org/10.1016/j.ymeth.2009.03.016">https://doi.org/10.1016/j.ymeth.2009.03.016</a>.</p>
</div>
<div id="ref-Ingolia2011-fc">
<p>Ingolia, N T, L F Lareau, and J S Weissman. 2011. “Ribosome Profiling of Mouse Embryonic Stem Cells Reveals the Complexity and Dynamics of Mammalian Proteomes.” <em>Cell</em> 147 (4): 789–802. <a href="https://doi.org/10.1016/j.cell.2011.10.002">https://doi.org/10.1016/j.cell.2011.10.002</a>.</p>
</div>
<div id="ref-Juntawong2014-ny">
<p>Juntawong, Piyada, Thomas Girke, Jérémie Bazin, and Julia Bailey-Serres. 2014. “Translational Dynamics Revealed by Genome-Wide Profiling of Ribosome Footprints in Arabidopsis.” <em>Proc. Natl. Acad. Sci. U. S. A.</em> 111 (1): E203–12. <a href="https://doi.org/10.1073/pnas.1317811111">https://doi.org/10.1073/pnas.1317811111</a>.</p>
</div>
<div id="ref-Juntawong2015-ru">
<p>Juntawong, Piyada, Maureen Hummel, Jeremie Bazin, and Julia Bailey-Serres. 2015. “Ribosome Profiling: A Tool for Quantitative Evaluation of Dynamics in mRNA Translation.” <em>Methods Mol. Biol.</em> 1284: 139–73. <a href="https://doi.org/10.1007/978-1-4939-2444-8%5C_7">https://doi.org/10.1007/978-1-4939-2444-8\_7</a>.</p>
</div>
<div id="ref-Kim2015-ve">
<p>Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. “HISAT: A Fast Spliced Aligner with Low Memory Requirements.” <em>Nat. Methods</em> 12 (4): 357–60.</p>
</div>
<div id="ref-Lawrence2013-kt">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-Robinson2010-uk">
<p>Robinson, M D, D J McCarthy, and G K Smyth. 2010. “EdgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.dcassol.com/">Daniela Cassol</a>, <a href="http://girke.bioinformatics.ucr.edu/">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '',
    indexName: '',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
